### PR TITLE
TagInput - Functional component for tag-like inputs

### PR DESCRIPTION
**Pull request by Gazook89**
# Description
This supersedes PR #3469.   This PR is just the very basic functionality of tags:  add, remove, update.  Subsequent PRs will layer on additional features/functionality in discrete chunks and descriptions are below.  Each of the additional features is a branch on the previous PR-- they build on each other.  As such, they can merged in order, or the last PR can be merged in and it will contain everything prior to it.  Here is the breakdown:

<details><summary>#3735 - Part 1 - Basic Functions</summary>
The PR removes the current StringArrayEditor.jsx entirely, and replaces it with tagInput.jsx which creates a functional component called TagInput.  Obviously it doesn't have to be limited to only "tags" as we called them in HB-- it can be used for the Author list as well and where you might want a *string array editor*.  I just prefer the shorter name for brevity.

- [x] add tags
- [x] remove tags individually
- [x] edit existing tags
- [x] save tags to brew metadata
</details>

<details><summary>[No PR yet] - Part 2 - Keyboard Navigation</summary>

Part 2 focuses on adding keyboard navigation to the tag input.  Things like using `Enter`, `,`, or `Tab` to submit a new value, or tabbing/shift+tabbing to move around the tags.

[Here is the branch on my remote](https://github.com/Gazook89/homebrewery/tree/TagInput-Part-2-Keyboard-Nav).  

</details>

<details><summary>[No PR yet] - Part 3 - Validations</summary>

Part 3 adds methods to validate the tag inputs so that we can control what values are input.  It differs from the original StringArrayEditor by having the validator functions centralized with other inputs in `validations.js`, rather than declaring the functions directly in props or as `valuePatterns`.  

[Here is the branch on my remote](https://github.com/Gazook89/homebrewery/tree/TagInput-Part-3-Validations). 

</details>

<details><summary>[No PR yet] - Part 4 - Uniqueness</summary>

Part 4 is basically a supplement to Part 3-- it adds a check to see if the new value is unique from the existing tags before submitting it.  By default, it must be unique, but via props the component could accept duplicate values.

[Here is the branch on my remote](https://github.com/Gazook89/homebrewery/tree/TagInput-Part-4-Uniqueness). 

</details>

## Quick Overview of TagInput
<details><summary>Click to expand</summary>

The component is called by the parent, MetadataEditor, which passes in an array of strings which represent the values of each tag (tag can be any data type, like Tags or Authors).  Additional props can be passed in, such as validation regexes, a boolean "unique?" prop, "non-editable" tags, etc-- though right now, most of those features aren't in this PR yet.

With the array of strings, TagInput creates a new array of objects.  Each object has two properties:  `value` and `editing`.  The latter is set to `false` by default.  Each tag is rendered as a list of divs that are "read only", or in the code, `renderReadTag()`.   These elements have a click event that changes the `editing` property of the tag to `true`, and converts them to "write" tags (`renderWriteTag()`), which are text inputs that allow you to edit the tag.  

TagInput passes an array of strings back up to MetadataEditor which the processes it with it's own `handleFieldChange` handler.

</details>

## Related Issues or Discussions

- Related discussion on previous PR #3469.   Most of the "good discussion" on the topic of tag inputs.

- [ ] fixes https://github.com/naturalcrit/homebrewery/issues/2333
- [ ] fixes https://github.com/naturalcrit/homebrewery/issues/3462

&nbsp;

- https://github.com/naturalcrit/homebrewery/issues/2335
- https://github.com/naturalcrit/homebrewery/issues/2357
- https://github.com/naturalcrit/homebrewery/issues/3166

## QA Instructions, Screenshots, Recordings

This is the visual design for THIS PR, Part 1:
image

For reference, here is a screenshot from Part 4:
image

### Reviewer Checklist

*Reviewers, refer to this list when testing features, or suggest new items *
- [ ] Verify new features are functional
	- [ ] add new tags, remove existing tags, edit existing tags
	- [ ] changes to tags are saved to brew, and are rendered on reload
	- [ ] loads existing tags (from brews that existed prior to this PR)
- [ ] Test for edge cases / try to break things
- [ ] Identify opportunities for simplification and refactoring
- [ ] Check for code legibility and appropriate comments

<details><summary>Copy this list</summary>

```
- [ ] Verify new features are functional
	- [ ] add new tags, remove existing tags, edit existing tags
	- [ ] changes to tags are saved to brew, and are rendered on reload
	- [ ] loads existing tags (from brews that existed prior to this PR)
- [ ] Test for edge cases / try to break things
- [ ] Identify opportunities for simplification and refactoring
- [ ] Check for code legibility and appropriate comments
```
</details

### CODE REVIEW

**Comment by 5e-Cleric** [comment]
Updated initial comment to actually have the list to copy

File: client/homebrew/editor/metadataEditor/metadataEditor.jsx [hunk 1]
```diff
@@ -363,12 +364,13 @@ const MetadataEditor = createClass({
...
 
-			<StringArrayEditor label='invited authors' valuePatterns={[/.+/]}
+			<TagInput label='invited authors' valuePatterns={[/.+/]}
 				validators={[(v)=>!this.props.metadata.authors?.includes(v)]}
 				placeholder='invite author' unique={true}
 				values={this.props.metadata.invitedAuthors}
 				notes={['Invited author usernames are case sensitive.', 'After adding an invited author, send them the edit link. There, they can choose to accept or decline the invitation.']}
-				onChange={(e)=>this.handleFieldChange('invitedAuthors', e)}/>
+				onChange={(e)=>this.handleFieldChange('invitedAuthors', e)}
+				/>
```
Comment (5e-Cleric) [inline client/homebrew/editor/metadataEditor/metadataEditor.jsx:373]:
```suggestion
				onChange={(e)=>this.handleFieldChange('invitedAuthors', e)}/>
```

File: client/homebrew/editor/tagInput/tagInput.jsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+	const handleInputKeyDown = ({ evt, value, index, options = {} }) => {
+		if (_.includes(['Enter', ','], evt.key)) {
+			evt.preventDefault();
+			submitTag(evt.target.value, value, index);
+			if (options.clear) {
+				setTemporaryValue('');
+			}
+		}
+	};
```
Comment (5e-Cleric) [inline client/homebrew/editor/tagInput/tagInput.jsx:28]:
```suggestion
	const handleInputKeyDown = ({ e, value, index, options = {} }) => {
		if (_.includes(['Enter', ','], e.key)) {
			e.preventDefault();
			submitTag(e.target.value, value, index);
			if (options.clear) {
				setTemporaryValue('');
			}
		}
	};
```

File: client/homebrew/editor/tagInput/tagInput.jsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+					{valueContext.map((context, index) => { return context.editing ? renderWriteTag(context, index) : renderReadTag(context, index); })}
+				</ul>
+
+				<input
+					type='text'
+					className='value'
+					placeholder={props.placeholder}
+					value={temporaryValue}
+					onChange={(e) => setTemporaryValue(e.target.value)}
+					onKeyDown={(evt) => handleInputKeyDown({ evt, value: null, options: { clear: true } })}
```
Comment (5e-Cleric) [inline client/homebrew/editor/tagInput/tagInput.jsx:98]:
```suggestion
					onKeyDown={(e) => handleInputKeyDown({ e, value: null, options: { clear: true } })}
```

File: client/homebrew/editor/tagInput/tagInput.jsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		handleChange(valueContext.map((context)=>context.value))
+	}, [valueContext])
+
+	const handleChange = (value)=>{
+		props.onChange({
+			target : { value }
+		})
+	};
+
+	const handleInputKeyDown = ({ evt, value, index, options = {} }) => {
```
Comment (calculuschild) [inline client/homebrew/editor/tagInput/tagInput.jsx:20]: Any special reason this function takes an object instead of individual parameters?

File: client/homebrew/editor/tagInput/tagInput.jsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
+require('./tagInput.less');
+const React = require('react');
+const { useState, useEffect } = React;
+const _     = require('lodash');
+
+const TagInput = ({ unique = true, values = [], ...props }) => {
+	const [temporaryValue, setTemporaryValue] = useState('');
```
Comment (calculuschild) [inline client/homebrew/editor/tagInput/tagInput.jsx:7]: I might name this `inputText` or something to convey what content this value represents.

File: client/homebrew/editor/tagInput/tagInput.jsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	const [valueContext, setValueContext] = useState(values.map((value) => ({ value, editing: false })));
```
Comment (calculuschild) [inline client/homebrew/editor/tagInput/tagInput.jsx:8]: I'm also not sure on this name. This refers to the current list of tags, right? Why not `tagList` or `tagArray` or something?

**Review by calculuschild — APPROVED** [review]
This seems functional and clean enough to me for a start. Just some variable naming suggestions.

Just nobody merge this yet until I can deploy the latest master branch live.

- [ ] Verify new features are functional
	- [x] add new tags, remove existing tags, edit existing tags
	- [x] changes to tags are saved to brew, and are rendered on reload
	- [ ] loads existing tags (from brews that existed prior to this PR)
- [ ] Test for edge cases / try to break things
- [x] Identify opportunities for simplification and refactoring
- [ x Check for code legibility and appropriate comments

File: client/homebrew/editor/tagInput/tagInput.jsx (hunk 4 above)
Comment (Gazook89) [reply client/homebrew/editor/tagInput/tagInput.jsx:20]: Not at computer right now but I believe it’s because it’s not always called with every parameter, so having  named properties helps

File: client/homebrew/editor/tagInput/tagInput.jsx (hunk 6 above)
Comment (Gazook89) [reply client/homebrew/editor/tagInput/tagInput.jsx:8]: It’s a holdover from the original string array editor. I can change it no problem

File: client/homebrew/editor/metadataEditor/metadataEditor.jsx (hunk 1 above)
Comment (Gazook89) [reply client/homebrew/editor/metadataEditor/metadataEditor.jsx:373]: I'm going to leave these as `evt` for now.  At the very end after each "stage" someone can flip these to `e` if really necessary.  I just prefer `evt` as it's easier to ctrl-f, though I understand that `e` is more commonly used throughout the project.

**Comment by Gazook89** [comment]
Going to keep track of a couple random things to fix, here, particularly if they don't fit neatly in upcoming related PRs:

- [ ] allow 'click anywhere' to *submit*(?) edited tag and drop focus.
- [ ] fix "must be unique" validation error if clicking into existing tag to edit, but then trying to submit without any changes  (can't change "chevy" to "chevy" since it is not "unique").
